core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -57,14 +58,16 @@ object RequestChannel extends Logging {
   }
 
   case class Request(processor: Int, connectionId: String, session: Session, private var buffer: ByteBuffer,
-                     startTimeMs: Long, listenerName: ListenerName, securityProtocol: SecurityProtocol) {
+                     startTimeNanos: Long, listenerName: ListenerName, securityProtocol: SecurityProtocol,
+                     networkThreadTime: AtomicLong) {
     // These need to be volatile because the readers are in the network thread and the writers are in the request
     // handler threads or the purgatory threads
-    @volatile var requestDequeueTimeMs = -1L
-    @volatile var apiLocalCompleteTimeMs = -1L
-    @volatile var responseCompleteTimeMs = -1L
-    @volatile var responseDequeueTimeMs = -1L
-    @volatile var apiRemoteCompleteTimeMs = -1L
+    @volatile var requestDequeueTimeNanos = -1L
+    @volatile var apiLocalCompleteTimeNanos = -1L
+    @volatile var responseCompleteTimeNanos = -1L
+    @volatile var responseDequeueTimeNanos = -1L
+    @volatile var apiRemoteCompleteTimeNanos = -1L
+    @volatile var recordNetworkThreadTimeCallback: Option[Long => Unit] = None
```

Comment:
Hmm... Network thread time needs to be accumulated against the (user, client-id) and needs to include the time for the sending the response. The callback avoids having to propagate (user, client-id).